bip-signmessage-redone.mediawiki

```diff
@@ -0,0 +1,156 @@
+<pre>
+  BIP: ????
+  Layer: Applications
+  Title: Generic Signed Message Format
+  Author: Karl-Johan Alm <karljohan-alm@garage.co.jp>
+  Comments-Summary: No comments yet.
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-????
+  Status: Draft
+  Type: Standards Track
+  Created: 2020-10-01
+  License: CC0-1.0
+  Replaces: 322
+</pre>
+
+== Abstract ==
+
+A standard for interoperable generic signed messages based on the Bitcoin Script format.
+
+== Background ==
+
+* Assume two actors, a prover <code>P</code> and a verifier <code>V</code>.
+* <code>P</code> wants to prove that they own the private key <code>k</code> associated with a given address <code>A</code> (which in turn is derived from the pubkey <code>kG</code>).
+* Let <code>V</code> generate a message <code>M</code> and hand this to <code>P</code>.
+* <code>P</code> generates a signature <code>S</code> by signing the message <code>M</code> using <code>k</code>. Given <code>S</code>, <code>V</code> can prove that <code>P</code> has the private key associated with <code>A</code>.
+
+The astute reader will notice that the above is missing a critical part, namely the pubkey <code>kG</code>, without which the verifier cannot actually verify the message. The current message signing standard solves this via a cryptographic trick, wherein the signature <code>S</code> above is a special "recoverable signature" type. Given the message <code>M</code> and the signature <code>S</code>, it is then possible to recover the pubkey <code>kG</code>. The system thus derives the address for the pubkey <code>kG</code>, and if it does not match <code>A</code>, the proof is deemed invalid.
+
+While this is a neat trick, it unnecessarily restricts and complicates the message signing mechanism; for instance, it is currently not possible to sign a message for a P2SH address, because there is no pubkey to recover from the resulting signature.
```

Comment:
Isn't it "there's no standard way to get from pubkey to the address" rather than "no pubkey to recover from the resulting signature" ?

Surely the signature could be of recoverable type, and if a prover and verifier use the same script template, the verifier can reconstruct the address and check (if all the other info such as other pubkeys are known to verifier and the template can be filled completely)

Reply:
With a p2pkh there is a pubkey hash, and with p2sh there isn't. Yes, there may be one or multiple keys embedded in the script hash, but there's no standardized way of determining that, as opposed to the p2pkh case.

Reply:
That was exactly my point in regards to the phrase "because there is no pubkey to recover from the resulting signature." -- It does not convey the correct meaning. The correct meaning is along the lines of "there's no standardized way of determining ...", "there's no standard way to get from recovered pubkey to the address"

Reply:
I see what you mean, changing.